Make bulk edit compatible with controlled vocabularies

ybnd opened the pull request:
## References
Fixes #9749 

## Description
This PR implements the solution proposed in #9749
- Consider the column header when determining whether the column is authority-controlled, instead of just the `[...]` suffix
- Be more strict about the authority value/confidence format in order to support `::` within the plain text value of a cell
    - If you provide an authority value, you must provide a confidence as well. Otherwise, the entire cell is considered to be the plain text value
    - Only the last two `::`s are considered to be authority/confidence separators, any previous `::`s are considered to be part of the plain text value

## Instructions for Reviewers
Confirm that this PR addresses the described issue without introducing others.
**Please consider whether it is sensible to alter existing CSV import behaviour in this way.**

### Testing
1. Set up a local DSpace REST instance on the `main` branch
2. Update the configuration to make the `dc.subject` authority-controlled wrt. the SRSC vocabulary, like in the IT configuration [here](https://github.com/ybnd/dspace/blob/bulk-edit-authority-issues/dspace-api/src/test/java/org/dspace/app/bulkedit/MetadataEditControlledVocabularyIT.java#L70-L82)
3. Deposit an Item with valid SRSC `dc.subject` values, including ones that are at least two "hierarchy levels deep"
    - These metadata will have `::` separators within the plain text values, as well as authority and confidence values
4. Export this Item using `metadata-export` and download the resulting CSV file
    - You will see that `::`s are used as both the hierarchy and authority separator
5. Make  afew (dummy) changes to the CSV file and re-import it using `metadata-import`
    - **The import will succeed, but the `dc.subject` values will be mangled**
6. Check out this PR's branch and rebuild DSpace
7. Run `metadata-import` on the same CSV
    - **The import will succeed, now the `dc.subject` values will be imported correctly**


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

ybnd committed:
Bulk edit is incompatible with controlled vocabularies

ybnd committed:
Support controlled vocabulary authority values

- Consider controlled vocabulary CSV columns authority-controlled if configured as such
- Support the default hierarchy delimiter `::` within authority-controlled CSV values
  - Authority must conform to `::([^:]+)::(\d+)`
  - If the last part is _not_ an integer, consider the entire string to be the metadata value instead
  - Requires us to drop support for "implicit confidence"

kshepherd commented:
Just a quick note to say I fully agree with the approach here, it fixes other unexpected behaviour when that `::` delimiter appears more than twice in a value string, too. Starting at the end and just parsing the last two segments looks good to me.

We may want to put some more explicit length testing on the parts array there, since if `parts.length` is < 3, the whole thing should probably just set the whole string as the value and return early since it must not be an authority value, and we'll avoid OOB errors.

kshepherd reviewed (changes requested):
@ybnd I just had one note:
We may want to put some more explicit length testing on the parts array there, since if parts.length is < 3, the new method should probably just set the whole string as the value and return early since it must not be an authority value, and we'll avoid OOB errors.

Does that seem reasonable?

ybnd committed:
Support plain values where the authority separator appears once

These can't be interpreted as valid values with authority (since we'd need a value, authority and confidence)
Therefore, the whole cell is used as the value instead.

ybnd commented:
@kshepherd sorry for the delay, that's a good idea indeed.

I've implemented your suggestions, and also made sure that the authority/confidence is explicitly "unset" in both cases. Not sure if that was strictly necessary, but it seems more consistent with [this part](https://github.com/DSpace/DSpace/pull/9753/files#diff-ae5c144a690d65c268a9b2eb456d483128bd7e3f3c5d228b1d0740ae449f6006R740).

kshepherd commented:
@ybnd it looks like a few controlled vocab ITs are failing now (not immediately clear to me why but it seems to maybe be discarding too much, or the tests themeselves just need updating?)

Commit:
129883: make the authorityControlled list no longer static so it re-initializes

ybnd committed:
Merge remote-tracking branch 'origin/main' into bulk-edit-authority-issues_9.0

ybnd commented:
@kshepherd 

@jensvannerum found that this was due to `authorityControlled` persisting between tests (sometimes it would work fine, sometimes not, depending on the order)
Since this is just some configuration getting read out before an import we're making it an instance property, let's see how the tests do now

kshepherd reviewed (approved):
+1 Looks great! Thanks @ybnd 
Gave this a practical test and the imported values are preserved as expected.
Merging immediately with 1 approval as this is a small change

lgeggleston commented:
It looks like there is renewed interest from the community in resolving this issue on 8.x: https://groups.google.com/g/dspace-tech/c/IyhGBF2yqTU/m/Y6lQKVCsAQAJ 

@ybnd, would you be willing to create a manual backport of this fix to 8.x possibly?

ybnd commented:
@lgeggleston yes certainly

This may also be a good opportunity for us to revisit https://github.com/DSpace/DSpace/issues/9896, I lost track of it
